### PR TITLE
Add a NAT and CGN rebind scenario

### LINKED ISSUE #26: create a scenario for NAT rebinding

### PR DESCRIPTION
**Pull request by larseggert**
Fixes #26.

The difference being that in the latter, client IP addresses also change.

### CODE REVIEW

File: sim/scenarios/nat/nat-error-model.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#include <unordered_map>
+
+using namespace ns3;
+
+// The NATErrorModel drops packets enumerated by the user. This model does
+// a simple packet count and drops the specified packets. Packet numbering
+// starts at 1.
+
+// A hash function used to hash a pair of any kind
+struct hash_pair {
```
Comment (marten-seemann) [inline sim/scenarios/nat/nat-error-model.h:16]: As far as I can see, we don't need this, right?
At least the code compiles without it.

File: sim/scenarios/nat/nat.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  cmd.AddValue("rebind_freq",
+               "NAT rebind frequency (e.g., 3s; 0 to disable repeated rebinds)",
+               rebind_freq);
+  cmd.AddValue("cgn", "act as CGN; change client IP address when rebinding",
+               cgn);
+  cmd.Parse(argc, argv);
+
+  NS_ABORT_MSG_IF(delay.length() == 0, "Missing parameter: delay");
+  NS_ABORT_MSG_IF(bandwidth.length() == 0, "Missing parameter: bandwidth");
+  NS_ABORT_MSG_IF(queue.length() == 0, "Missing parameter: queue");
```
Comment (marten-seemann) [inline sim/scenarios/nat/nat.cc:38]: Can you add aborts for the other variables as well? As far as I can tell, `first_rebind` should be required.

File: sim/scenarios/nat/nat-error-model.cc [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    rev[old_port] = 0;
+    cout << Simulator::Now().GetSeconds() << "s: " << (do_cgn ? "CGN" : "NAT")
+         << " rebinding: " << old_nat << ":" << old_port << " -> " << nat << ":"
+         << b.second << endl;
+  }
+
+  // for (auto &b : fwd)
+  //   cout << "fwd: " << b.first << " -> " << b.second << endl;
+  // for (auto &b : rev)
+  //   cout << "rev: " << b.first << " -> " << b.second << endl;
```
Comment (marten-seemann) [inline sim/scenarios/nat/nat-error-model.cc:50]: Nit: remove commented code.

File: sim/scenarios/nat/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+IP address also changes and not just the port.) All flows are rebound to new
+source ports (and possibly IP a new IP address) every time the timer expires.
+
+This scenario has the following configurable properties:
+
+* `--first_rebind`: The time of the first NAT/CGN rebind event. Zero to disable.
+  For example, `--first_rebind=1s`.
+
+* `--rebind_freq`: The frequency for subsequent NAT/CGN rebinds after the first
+  event. For example, `--rebind_freq=3s`.
```
Comment (marten-seemann) [inline sim/scenarios/nat/README.md:16]: What happens for 0? No rebinding after the first one?
(Maybe point out that this parameter is optional if that's the case).

File: sim/scenarios/nat/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+scenario and simulates a NAT that is frequently changing the source addresses of
+all client connections. This scenario can be used to test whether QUIC
+implementations correctly handle NAT and CGN rebindings (in the latter case, the
+IP address also changes and not just the port.) All flows are rebound to new
+source ports (and possibly IP a new IP address) every time the timer expires.
+
+This scenario has the following configurable properties:
+
+* `--first_rebind`: The time of the first NAT/CGN rebind event. Zero to disable.
+  For example, `--first_rebind=1s`.
```
Comment (marten-seemann) [inline sim/scenarios/nat/README.md:13]: Why treat 0 special here? Disabling doesn't make a lot of sense here, does it? If people don't want NAT rebinding, they can just use simple-p2p.

File: sim/scenarios/helper/quic-point-to-point-helper.h [hunk 6]
```diff
@@ -2,6 +2,7 @@
 #define QUIC_POINT_TO_POINT_HELPER_H
 
 #include "ns3/point-to-point-module.h"
+#include "ns3/string.h"
```
Comment (marten-seemann) [inline sim/scenarios/helper/quic-point-to-point-helper.h:5]: Leftover?

Comment (larseggert) [reply sim/scenarios/helper/quic-point-to-point-helper.h:5]: probably

File: sim/scenarios/nat/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# NAT Rebinding
```
Comment (janaiyengar) [inline sim/scenarios/nat/README.md:1]: My recommendation here follows what I wrote in the interop runner PR -- call this Port Rebinding. NAT rebinding is too generic and doesn't capture precisely what this scenario does.

File: sim/scenarios/nat/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+* `--first_rebind`: The time of the first NAT/CGN rebind event. Zero to disable.
+  For example, `--first_rebind=1s`.
+
+* `--rebind_freq`: The frequency for subsequent NAT/CGN rebinds after the first
+  event. For example, `--rebind_freq=3s`.
+
+* `--cgn`: Whether the NAT acts as a CGN and also rewrites addresses. Default is
+  `false`. For example, `--cgn` enables CGN mode.
+
```
Comment (janaiyengar) [inline sim/scenarios/nat/README.md:20]: I would call this the --ip option.

File: sim/scenarios/nat/README.md (hunk 5 above)
Comment (janaiyengar) [reply sim/scenarios/nat/README.md:13]: I would require this to be > 0.

**Review by janaiyengar — COMMENTED** [review]
Thanks for working on this, @larseggert -- this is great. A few comments inline.

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**